source/reporting/man/man1/arcorun.1
'\" t
.\"___INFO__MARK_BEGIN__
.\"
.\"The Contents of this file are made available subject to the terms of
.\"the Sun Industry Standards Source License Version 1.2
.\"Sun Microsystems Inc., March, 2001
.\"
.\"
.\"Sun Industry Standards Source License Version 1.2
.\"=================================================
.\"The contents of this file are subject to the Sun Industry Standards
.\"Source License Version 1.2 (the "License"); You may not use this file
.\"except in compliance with the License. You may obtain a copy of the
.\"License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
.\"
.\"Software provided under this License is provided on an "AS IS" basis,
.\"WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
.\"WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
.\"MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
.\"See the License for the specific provisions governing your rights and
.\"obligations concerning the Software.
.\"
.\"The Initial Developer of the Original Code is: Sun Microsystems, Inc.
.\"
.\"Copyright: 2001 by Sun Microsystems, Inc.
.\"
.\"All Rights Reserved.
.\"
.\"#######################################################################
.\"
.\"___INFO__MARK_END__
.\"
.\" $RCSfile: arcorun.1,v $ 
.\"
.\"
.\" Some handy macro definitions [from Tom Christensen's man(1) manual page].
.\"
.de SB		\" small and bold
.if !"\\$1"" \\s-2\\fB\&\\$1\\s0\\fR\\$2 \\$3 \\$4 \\$5
..
.\"
.de T		\" switch to typewriter font
.ft CW		\" probably want CW if you don't have TA font
..
.\"
.de TY		\" put $1 in typewriter font
.if t .T
.if n ``\c
\\$1\c
.if t .ft P
.if n \&''\c
\\$2
..
.\"
.de M		\" man page reference
\\fI\\$1\\fR\\|(\\$2)\\$3
..
.TH ARCORUN 1 "$Date: 2008/03/31 15:58:23 $" "xxRELxx" "xxQS_NAMExx User Commands"
.SH NAME
arcorun \- execute queries and save the results in XML, CSV, HTML or PDF format
.\"
.\"
.SH SYNTAX
.B arcorun
[
.B -c file
] [
.B -d debug_level
] [
.B -f format
] [
.B -help
] [
.B -l
] [
.B -lb name=value
]* [
.B -lbfile file
] [
.B -n result_name
] [
.B -o output_file
] [
.B -v
] [
.B -?
] [
.B query_name
]
.\"
.SH DESCRIPTION
.I arcorun
allows the user to list and execute ARCo queries from the command line.  The
output format of the queries can be set to XML (default), CSV, PDF or HTML.
Values for late binding parameters can also be set.  In order for
.I arcorun
to be used, it must have access to the ARCo spooling directory (default:
/var/spool/arco).  Because of this,
.I arcorun
must be run from a host from which the spooling directory is accessible.
.PP
.\"
.\"
.SH OPTIONS
.\"
.IP "\fB\-c file\fP"
Sets the path to the ARCo configuration file.  If not set, the default path is
/$SGE_ROOT/$SGE_CELL/arco/reporting/config.xml
.\"
.IP "\fB\-d debug_level\fP"
Sets the amount of information
.I arcorun
will generate.  Options in order of
increasing information are: \fBWARNING\fP, \fBINFO\fP, \fBFINE\fP
.\"
.IP "\fB\-f format\fP"
Sets the output format for the query results.  Possible values of \fBfile\fP
are "xml", "csv", "pdf", and "html".  "xml" is the default.  Query results
stored in XML format can be read by the ARCo console for graphical viewing.
"csv" outputs values in CSV format, i.e. entries separated by commas and
records separated by lines.  "pdf" produces a PDF document.  The output should
either be redirected to a file, or the \fI\-o\fP option should be used to store
the output in a file.  "html" produces a set of HTML files and resources which
contain the query results.  The \fI\-o\fP option \fBmust\fP be used with
\fI\-f html\fP.
.\"
.IP "\fB\-help\fP"
Prints usage information and exits.
.\"
.IP "\fB\-l\fP"
Lists the available queries.
.\"
.IP "\fB\-lb name=value\fP"
Sets the late-binding parameter, \fBname\fP, to the value, \fBvalue\fP.
.\"
.IP "\fB\-lbfile file\fP"
Specifies a file to be read to specify late-binding parameters.  Each
late-binding parameter should be on its own line, and should be in the form:
name=value.
.\"
.IP "\fB\-n result_name\fP"
Sets the name given to the results.  If this option is not specified
the results take the name of the given query.
.\"
.IP "\fB\-o output_file\fP"
Stores the query results in the specified file.  If not set, query results are
output to stdout.  In the case of \fI\-f html\fP, \fBoutput_file\fP must be the
path to a directory.  In the case where XML output is to be viewed in the ARCo
console, the output must be stored in the results subdirectory of the ARCo
spooling directory, normally /var/spool/arco/results.
.\"
.IP "\fB\-v\fP"
Displays the
.I arcorun
version information.
.\"
.IP "\fB\-?\fP"
Same as \fI\-help\fP
.\"
.\"
.SH "OUTPUT FORMATS"
The output format is determined by the \fI\-f\fP option.  If this option is not
set, the output format defaults to XML.  See the \fI\-f\fP option description
for more information.
.\"
.\"
.SH "EXAMPLES"
.SS "\fBRunning a Query\fP"
A query is run by simply invoking the arcorun command with the name of the query
as the argument:

.nf
% $SGE_ROOT/$SGE_CELL/arco/reporting/arcorun Statistics
.fi

If a query name contains whitespaces you have to put double-quotes around the
query name:

.nf
% $SGE_ROOT/$SGE_CELL/arco/reporting/arcorun "Host Load"
.fi

.SS "\fBLate Bindings\fP"

The Query "Job Log" defines the late-binding parameter, job_number:

.nf
.RS
% $SGE_ROOT/$SGE_CELL/arco/reporting/arcorun \\
            -lb job_number=2311 "Job Log"
.RE
.fi

Alternatively, with the -lbfile option:

.nf
.RS
% cat lbfile.txt
job_number=2311
% $SGE_ROOT/$SGE_CELL/arco/reporting/arcorun \\
            -lbfile lbfile.txt "Job Log"
.RE
.fi

.SS "\fBViewing generated reports\fP"
Results whith are stored in XML output format can be viewed in the in
ARCo web console.  Execute the query with the arcorun util and use the
\fI\-o output_file\fP flag to store the result in the results subdirectory of
the ARCo spool directory (/var/spool/arco).

The following script can be used to generate a monthly report of compute host
load on the grid:

.nf
.RS
#!/bin/sh

SGE_ROOT=/tools/sge
SGE_CELL=default

SHORT_MONTH=`date +%m`
LONG_MONTH=`date +%b`
YEAR=`date +%y`
RESULT_FILE=dep_acct_${SHORT_MONTH}.xml
RESULT_NAME="Host Load (${LONG_MONTH} ${YEAR})"
QUERY_NAME="Host Load"

$SGE_ROOT/$SGE_CELL/arco/reporting/arcorun \\
          -o /tmp/$RESULT_FILE \\
          -n "$RESULT_NAME" \\
          "$QUERY_NAME"
if [ $? = 0 -a -f /tmp/$RESULT_FILE ]; then
  mv /tmp/$RESULT_FILE /var/spool/arco/results/$RESULT_FILE
fi
.RE
.fi

The query "Host Load" should be previously defined and should provide a
report on host load for the past month.  When executed, the results of this
script will be visible in the \fBResult List\fP of the ARCo web module.  A
restart of the Java Web Console is not necessary; simply reload the Overview
page.

By adding the following script to the cron tables, the above monthly report will
be automatically created:

.nf
% crontab -e
...
# run at 2:00am on the first of every the host load report
0 2 1 * * /opt/sge/default/arco/reporting/monthly_report.sh
...
.fi

This cron job should normally be run as the user as which the ARCo web console
runs (default: noaccess), because the generated results can only be deleted if
the web console, running as that user, has write permissions to the generated
result file(s).
.\"
.\"
.SH FILES
.nf
.ta \w'<xxqs_name_sxx_root>/     'u
\fI/var/spool/arco/results\fP
	The ARCo results directory
.fi
.\"
.\"
.SH "COPYRIGHT"
See
.M xxqs_name_sxx_intro 1
for a full statement of rights and permissions.


Changelog
RD-2008-03-04-0: CVS:       Created V62beta_RC1_TAG
---------------------------------------------------

JO-2008-03-31-0: Enhancem.:   changed messages according to Janice advice
                 Review:      AP

JO-2008-03-27-0: Enhancem.:   Fixed typos, consistency changes, changed messages according to Janice advice
                 Review:      AP

PJ-2008-03-27-0: Bugfix:      Latebindings for Advanced Queries does not work
                 Bugster:     6677885
                 Issue:       2529
                 Bugfix:      Repaired object synchronization in ArcoServlet
                 Bugfix:      Fixed online help for advance query latebindings
                 Review:      JO 

PJ-2008-03-20-0: Bugfix:      Advanced Query with wild card character * does not produce correct results.
                 Enhancem.:   The SQLQueryResultsets is following the names from the database row set meta data.
                 Enhancem.:   The lifecycle of the SQLQueryResultsets is optimised. Passivate is being called in finally
                 Enhancem.:   Unused statements and results pro ArcoDbConnection is removed
                 Enhancem.:   The arco.xsd is reorganised. The related parts are moved together. 
                 Bugster:     6511009
                 Review:      LP

LP-2008-03-18-1: Bugfix:      Fixed typo from LP-2008-03-18-0
                 Review:      JO

LP-2008-03-18-0: Enhancem.:   Added useful message for users when startup fails 
                              with SMF
                 Review:      JO

AP-2008-03-17-0: Bugfix:      Fix of incorrect usage summary for larger databases.
                 Enhancem.:   ARCo reporting installation - get all required SWC
                              information for installation.
                 Review:      JO, LP

PJ-2008-03-03-0: Bugfix:      Validation of the Filter List in Simple Query builder fails
                 Bugfix:      Row Limit in ARCo Simple Query builder cannot be empty
                 Bugster:     6670264, 6661461 
                 Review:      JO

PJ-2008-02-29-0: Bugfix:      Row Limit in ARCo Simple Query builder cannot be empty
                 Bugster:     6661461
                 Review:      JO

PJ-2008-02-28-0: Bugfix:      Sort order and row limit cannot be specified together in ARCo Simple Query builder
                 Bugster:     6661470
                 Review:      JO

JO-2008-02-28-0: Enhancem.:   Enhanced reporting test SQLGeneratotTest
                 Review:      PJ

JO-2008-02-27-0: Bugfix:      Fixed commit JO-2008-02-22-0 to work with arcorun
                 Review:      PJ

JO-2008-02-22-0: Bugfix:      In SJWC on Oracle dates appear truncated to just MM/DD/YYYY
                 Bugster:     6661500
                 Issue:       2503
                 Review:      AP

AP-2008-02-20-0: Bugfix:      ARCo reporting module installation script is broken 
                              on Red Hat Enterprise Linux 4 Update 4
                 Bugster:     6663741
                 Issue:       2498
                 Review:      JG

JO-2008-02-15-0: Bugfix:      Batch failure on Oracle database.
                 Bugster:     6659300
                 Review:      PJ

JO-2008-02-11-0: Bugfix:      Queries Average Job Turnaround Time and Average Job Wait Time irresponsive on MySQL
                 Bugster:     6661473
                 Review:      AP

AP-2008-02-06-0: Bugfix:      Correct the field name in dbwriter.xml for mysql 
                              to ju_exit_status

RD-2008-02-04-0: CVS:         Created V62alpha1_TAG
---------------------------------------------------

LP-2008-02-01-0: Cleanup:     Cleanup to use new GetCell functionality (detection)
                 Review:      PJ

JO-2008-02-01-0: Cleanup:     got rid of 5.3 specific code
                              got rid of word Please in prompts
                              changes for consistency
                              other minor cosmetic changes
                 Review:      PJ    

PJ-2008-01-31-0: Bugfix:      ORDER BY clause ignored in Advanced Query
                 Bugster:     6598005 
                 Issue:       2354
                 Review:      JO

LP-2008-01-25-0: Enhancem.:   Improvements to SMF support
                              Cleanup for java 1.5

PJ-2008-01-25-0: Bugfix:      Sort on table column throws exception 
                              if explicit SORT specified in 
                              SimpleQuery Sort on table column
                 Bugster:     6549694
                 Issue:       2344
                 Review:      LP

AP-2008-01-22-0: Bugfix:      deletion rule for PostgreSQL incorrect for deletion 
                              of sge_share_log
                 Bugster:     6653090
                 Issue:       2468
                 Review:      JO

PJ-2008-01-16-0: Bugfix:      Row Limit in Simple Query uses wrong syntax
                 Bugster:     6640762
                 Issue:       2444
                 Review:      JO

LP-2008-01-16-0: Bugfix:      PID file is not automatically removed

PJ-2008-01-16-0: Enhancem.:   Change ClusterIndex to ClusterName, ready for unique
                              SGE_CLUSTER_NAME
                 Review:      JO

AP-2008-01-16-0: Bugfix:      Wrong variable for calculating daily host values from 
                              hourly ones
                              DBW derived rules and reporting queries that count 
                              jobs need to be updated - updating views view_accounting, 
                              view_job_times, view_jobs_completed

source/reporting/src/com/sun/grid/arco/ArcoRun.java
/*___INFO__MARK_BEGIN__*/
/*************************************************************************
 *
 *  The Contents of this file are made available subject to the terms of
 *  the Sun Industry Standards Source License Version 1.2
 *
 *  Sun Microsystems Inc., March, 2001
 *
 *
 *  Sun Industry Standards Source License Version 1.2
 *  =================================================
 *  The contents of this file are subject to the Sun Industry Standards
 *  Source License Version 1.2 (the "License"); You may not use this file
 *  except in compliance with the License. You may obtain a copy of the
 *  License at http://gridengine.sunsource.net/Gridengine_SISSL_license.html
 *
 *  Software provided under this License is provided on an "AS IS" basis,
 *  WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING,
 *  WITHOUT LIMITATION, WARRANTIES THAT THE SOFTWARE IS FREE OF DEFECTS,
 *  MERCHANTABLE, FIT FOR A PARTICULAR PURPOSE, OR NON-INFRINGING.
 *  See the License for the specific provisions governing your rights and
 *  obligations concerning the Software.
 *
 *   The Initial Developer of the Original Code is: Sun Microsystems, Inc.
 *
 *   Copyright: 2001 by Sun Microsystems, Inc.
 *
 *   All Rights Reserved.
 *
 ************************************************************************/
/*___INFO__MARK_END__*/
package com.sun.grid.arco;

import com.sun.grid.logging.SGELog;
import com.sun.grid.arco.sql.ArcoDbConnectionPool;
import java.io.*;
import java.util.Properties;
import java.util.logging.ConsoleHandler;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.xml.transform.TransformerException;
import com.sun.grid.arco.model.NamedObject;
import com.sun.grid.arco.model.Query;
import com.sun.grid.arco.model.StorageType;
import com.sun.grid.arco.sql.SQLQueryResult;
import com.sun.grid.logging.SGEFormatter;
import java.util.Collection;
import java.util.Enumeration;
import java.util.Iterator;
import java.util.Locale;

/**
 * Command line to for ARCo
 * with this class ARCo queries 
 */
public class ArcoRun
 {
   /** In this mode arcorun will export the result of a query */
   public static final int MODE_EXPORT = 0;
   /** In this mode arcorun will list the names of all available queries */
   public static final int MODE_LIST   = 1;
   /** In this mode arcorun will print a usage message */
   public static final int MODE_HELP   = 2;
   /** In this mode arcorun will print the version string */
   public static final int MODE_VERSION = 3;
   /** the default log level */
   public static final Level DEFAULT_LOG_LEVEL        =  Level.INFO;
   /** export the query result as xml */
   public static final int OUTPUT_FORMAT_XML  = 0;
   /** export the query result as csv */
   public static final int OUTPUT_FORMAT_CSV  = 1;
   /** export the query result as pdf */
   public static final int OUTPUT_FORMAT_PDF  = 2;
   /** export the query result as html */
   public static final int OUTPUT_FORMAT_HTML = 3;
   
   
   /** Logger of this application */
   private Logger logger;
   /** The log hander for this appliation */
   private ConsoleHandler consoleHandler;   
   /** The log level for this application */
   private Level      logLevel = Level.INFO;      
   /** The arco configuration file */
   private File       arcoConfigFile;
   /** the output file */
   private File       outputFile = null;
   /** the output format */
   private int        outputFormat = OUTPUT_FORMAT_XML;
   /** the latebinging properties */
   private Properties latebindings = new Properties();
   /** name of the query */
   private String     queryName;
   /** name of the result, if null it is taken from the query */
   private String     resultName;
   /** the mode of the application */
   private int mode = MODE_EXPORT;
   /** directory where arco is installed ( System property arco.home) */
   private File arcoHome;

   /**
    *  Run the application 
    */
   public void run() throws IOException, TransformerException, ArcoException {
      switch( mode ) {
         case MODE_EXPORT:    export(); break;
         case MODE_HELP:      usage();  break;
         case MODE_LIST:      list();   break;
         case MODE_VERSION:   System.out.println( getVersion() );  
            break;
         default:
            throw new IllegalArgumentException("Unknown mode " + mode );            
      }
   }
   
   /**
    *  get the version string of this application
    *  @return the version string in the form "<product> <version> - arcorun"
    */
   public static String getVersion() {
      return ArcoVersion.PLUGIN_NAME + " - " +
             ArcoVersion.VERSION + " - arcorun";
   }
   
   private void initConfigFile() {
      // Setup the configuration file
      String cf = System.getProperty("arco.config");
      arcoConfigFile = new File(cf + File.separator +"config.xml");      
   }
   
   
   /**
    *  list all available queries 
    */
   private void list() throws ArcoException {

     ArcoDbConnectionPool dbConnections = ArcoDbConnectionPool.getInstance();
     
     dbConnections.setConfigurationFile( arcoConfigFile );
     
     StorageType storage = dbConnections.getConfig().getStorage();

     File storageDir = new File( storage.getRoot() );                       
     File queryDir = new File( storageDir, storage.getQueries() );
     
     QueryManager queryManager = new QueryManager(queryDir);
     
     NamedObject [] queries = queryManager.getAvailableObjects();
     
     for( int i = 0; i < queries.length; i++ ) {
        System.out.println( queries[i].getName() );
     }
   }
   
   /**
    * Loads and executes a query. It saves the result as xml and 
    * exports it to cvs.
    * @throws ArcoException   on any expected error
    * @throws IOException     if the result could not be written due to an I/O Error
    * @throws javax.xml.transform.TransformerException  if a transformation with a XSL stylesheet failed
    */
   public void export() throws IOException, TransformerException, ArcoException {
     SGELog.entering( getClass(), "run" );
     
     ArcoDbConnectionPool dbConnections = ArcoDbConnectionPool.getInstance();

     dbConnections.setConfigurationFile( arcoConfigFile );
     
     StorageType storage = dbConnections.getConfig().getStorage();

     File storageDir = new File( storage.getRoot() );                       
     File queryDir = new File( storageDir, storage.getQueries() );
     File resultDir = new File( storageDir, storage.getResults() );
     
     QueryManager queryManager = new QueryManager( queryDir );     
     
     Query query = queryManager.getQueryByName(queryName);
     
     if(query == null) {
        throw new ArcoException("ArcoRun.queryNotFound", new Object [] { queryName });
     }
     QueryResult queryResult = new SQLQueryResult(query,dbConnections);
     
     if(queryResult.hasLateBinding() ) {
        Enumeration propNameEnum = latebindings.propertyNames();
        String propName = null;
        while( propNameEnum.hasMoreElements() ) {
           propName = (String)propNameEnum.nextElement();
           queryResult.setLateBinding(propName,  latebindings.getProperty(propName ) );
        }
        
        Collection lateBindingNames = queryResult.getLateBindingsNames();
        Iterator iter = lateBindingNames.iterator();
        while(iter.hasNext()) {
           propName = (String)iter.next();
           if( queryResult.getLateBinding(propName) == null ) {
              throw new ArcoException("ArcoRun.missingLatebinding", 
                                      new Object[] { propName } );
           }
        }
     }
     
     queryResult.execute();
          
     if( queryResult.getRowCount() > 0 ) {
        
        if(this.resultName != null ) {
           queryResult.getQuery().setName(resultName);
        }
        
        ResultExportManager exportManager = new ResultExportManager( arcoHome );
        
        switch( outputFormat ) {
           case OUTPUT_FORMAT_XML:
              exportManager.export( ResultExportManager.TYPE_XML, queryResult,createOutputStream(), Locale.getDefault() );
              break;
           case OUTPUT_FORMAT_CSV:
              exportManager.export( ResultExportManager.TYPE_CSV,  queryResult,createOutputStream(), Locale.getDefault() );
              break;
           case OUTPUT_FORMAT_PDF:
              exportManager.export( ResultExportManager.TYPE_PDF,  queryResult,createOutputStream(), Locale.getDefault() );
              break;
           case OUTPUT_FORMAT_HTML:
              exportManager.export( ResultExportManager.TYPE_HTML, queryResult, outputFile, Locale.getDefault() );
              break;
           default:
              throw new IllegalStateException("unknown output format " + outputFormat );
        }
     } else {
        SGELog.warning( "ArcoRun.noResult" );
     }
     dbConnections.releaseConnections();
     SGELog.exiting( getClass(), "run" );
   }

   /**
    *  create the output stream where the result of the export will be stored
    *  @return the output stream
    *  @throws com.sun.grid.arco.ArcoException if the stream could not be created
    */
   private OutputStream createOutputStream() throws ArcoException {
       try {
          if( this.outputFile == null ) {
             return System.out;
          } else {
             return new FileOutputStream( outputFile );
          }
       } catch( IOException ioe ) {
          throw new ArcoException("ArcoRun.outputFileOpenError", ioe, 
                              new Object[] { ioe.getMessage() } );
       }
   }
      
   /** 
    *   intializes the logging 
    */   
   private void initLogging() {
      
      int columns [] = {
         SGEFormatter.COL_LEVEL_LONG, SGEFormatter.COL_MESSAGE
      };
      
      logger = Logger.getLogger( "arcoLogger", "com.sun.grid.arco.web.arcomodule.Resources" );
       
      consoleHandler = new ConsoleHandler();
      SGEFormatter  formatter = new SGEFormatter( "arcorun", false, columns );
      
      formatter.setDelimiter( ": " );
       
      consoleHandler.setFormatter( formatter );
      logger.addHandler( consoleHandler );
      logger.setUseParentHandlers( false );
       
      SGELog.init( logger );
   }
   

   
   /**
    * @param args the command line arguments
    */
   public static void main(String args[]) {
      
      ArcoRun app = new ArcoRun();      
            
      //set the default config file from SWC property
      app.initConfigFile();
      app.initLogging();
      int exit = 0;
      try {
         
         try {
            app.setArguments(args);        
         } catch( ArcoException ae ) {
            SGELog.severe_p(ae, ae.getMessage(), ae.getParameter() );
            app.usage();         
            exit = -1;
         }
         if( exit == 0 ) {
            app.run();
         }
         
      } catch( ArcoException ae ) {         
         SGELog.severe_p( ae, ae.getMessage(), ae.getParameter() );
         exit = -2;
      } catch (Exception ex) {         
         SGELog.severe( ex, "Unexpected Error: {0}", ex.getMessage() );
         exit = -3;
      }
      System.exit( exit );
   }
   
   
   /** 
    * Set the command line arguments
    * @param args list of commandline parameters
    * @throws ArcoException if the argument are not valid
    */   
   public void setArguments(String[] args) throws ArcoException {
       if (args.length < 1) {
          throw new ArcoException("ArcoRun.illegalArgumentCount");
       }
       
     String arcoHomeStr =  System.getProperty( "arco.home" );
     
     if( arcoHomeStr == null ) {
        throw new ArcoException("ArcoRun.missingArcoHome");
     }
     
     arcoHome = new File( arcoHomeStr );
       
     if( !arcoHome.isDirectory() ) {
        throw new ArcoException("ArcoRun.arcoHomeIsNoDirectory", new Object[] { arcoHome } );
     }
       // parse the options
     
     for( int i = 0; i < args.length; i++ ) {
        if( args[i].equals( "-c" ) ) {
           i++;
           if( i >= args.length ) {
              throw new ArcoException("ArcoRun.arcoConfFileisMissing");
           }
           arcoConfigFile = new File( args[i] );
        } else if ( args[i].equals( "-d" ) ) {
           i++;
           if( i >= args.length ) {
              throw new ArcoException("ArcoRun.debugLevelIsMissing");
           }
           try {
              logLevel = Level.parse( args[i] );
              SGELog.info( "set debug level to {0}", logLevel.toString() );
              logger.setLevel( logLevel );
              consoleHandler.setLevel( logLevel );
           } catch( IllegalArgumentException ile ) {
              throw new ArcoException( "ArcoRun.invalidLogLevel", new Object[] { args[i] } );
           }
        } else if ( args[i].equals( "-o" ) ) {
           i++;
           if( i >= args.length ) {
              throw new ArcoException( "ArcoRun.outputFileMissing");
           }
           outputFile = new File( args[i] );
        } else if ( args[i].equals( "-f" ) ) {
           i++;
           if( i >= args.length ) {
              throw new ArcoException( "ArcoRun.outputFormatMissing");
           }
           
           outputFormat = strToOutputFormat( args[i] );
           if( outputFormat < 0 ) {
              throw new ArcoException( "ArcoRun.unknownOutputFormat", new Object[] {args[i]} );
           }
        } else if( args[i].equals( "-lb" ) ) {
           i++;
           if( i >= args.length ) {
              throw new ArcoException("ArcoRun.missingLateBindingParameter");
           }
           int index = args[i].indexOf( '=' );
           if( index <= 0 ) {
              throw new ArcoException("ArcoRun.invalidLateBindingParameter", new Object[] { args[i] } );
           }
           String name  = args[i].substring( 0, index ).trim();
           String value = args[i].substring( index + 1 ,args[i].length() ).trim();
           
           SGELog.fine( "late binding {0} = {1}", name, value );
           latebindings.setProperty( name, value );
        } else if ( args[i].equals( "-lbfile" ) ) {
           
           i++;
           if( i >= args.length ) {
              throw new ArcoException("ArcoRun.missingLateBindingFile");
           }
           
           Properties props = new Properties();
           
           try {
              FileInputStream fin = new FileInputStream( args[i] );
              props.load( fin );
              
              Enumeration names = props.propertyNames();
              String name = null;
              String value = null;
              while( names.hasMoreElements() ) {
                 name = (String)names.nextElement();
                 value = props.getProperty( name );
                 SGELog.fine( "late binding {0} = {1}", name, value );
                 latebindings.setProperty( name, value );
              }
           } catch( IOException ioe ) {
              throw new ArcoException("ArcoRun.latebindingIOError", ioe, new Object[] { args[i], ioe.getMessage() } );
           }
        } else if ( args[i].equals("-n")) {
           i++;
           if(i >= args.length ) {
              throw new ArcoException("ArcoRun.missingResultName");
           }
           this.resultName = args[i];
           SGELog.fine("resultName is {0}", resultName);
        } else if ( args[i].equals( "-l" ) ) {
           mode = MODE_LIST;
        } else if ( args[i].equals( "-help" ) || args[i].equals("-?") ) {
           mode = MODE_HELP;
        } else if ( args[i].equals( "-v" ) ) {
           mode = MODE_VERSION;
        } else if( i == args.length - 1 ) {
           queryName = args[i];
        } else {
           throw new ArcoException("ArcoRun.unknownOption", new Object[] { args[i] } );
        }
     }
     
     if( mode == MODE_EXPORT ) {
        
        if( queryName == null ) {
           throw new ArcoException("ArcoRun.missingQueryName");
        }
        
        if( outputFormat == OUTPUT_FORMAT_HTML ) {
           if( outputFile == null  ) {
              throw new ArcoException("ArcoRun.missingOutputForHTML" );
           } else if ( !outputFile.isDirectory() ) {
              throw new ArcoException("ArcoRun.invalidOutputForHTML", new Object[] { outputFile } );
           }
        }
     }
     SGELog.fine( "parameters list is ok" );
   }

   /**
    * convert a string into a output format constant
    * @param str   the string
    * @return the output format constant or <code>-1</code>
    */
   public static int strToOutputFormat( String str ) {
      if( str.equals( "xml" ) ) {
          return OUTPUT_FORMAT_XML;
      } else if ( str.equals( "csv" ) ) {
         return OUTPUT_FORMAT_CSV;         
      } else if ( str.equals( "pdf" ) ) {
         return OUTPUT_FORMAT_PDF;
      } else if ( str.equals( "html" ) ) {
         return OUTPUT_FORMAT_HTML;
      } else {
         return -1;
      }
   }
   
   
   /** 
    * Print the usage of htis tool to standard err.
    */
   public void usage() {
     System.err.println( getVersion() );
     System.err.println( "arcorun  [-c <file>] [-d <debuglevel>]" );
     System.err.println( "         [-l] [-help] [-?] [-v] [-o <output file>] [-n <result name>]");
     System.err.println( "         [-f <format>] (-lb name=value)* [-lbfile <file>]  <query name> )" );
     System.err.println( );
     System.err.println( "       -l                  list all available query names");
     System.err.println( "       -help | -?          print this help message");
     System.err.println( "       -v                  print version");
     System.err.println( "       -c <file>           path to set configuration file of arco" );
     System.err.println( "                           (default: " + "$SGE_ROOT/$SGE_CELL/arco/reporting/config.xml)" );
     System.err.println( "       -d <level>          debug level for arco run (FINE INFO WARNING)");
     System.err.println( "       -o <output file>    path of the file where the result of the query will be stored");
     System.err.println( "                           (default stout)");
     System.err.println( "       -f <format>         format of the output (xml|csv|pdf|html). if the output format" );
     System.err.println( "                           is html the output file must be a directory.");
     System.err.println( "                           (default xml)");
     System.err.println( "       -lb <name>=<value>  specify the a late binding parameter for the query");
     System.err.println( "       -lbfile <file>      where late binding parameters are specified");
     System.err.println( "       -n <result name>    name of the result, if this option is not specified");
     System.err.println( "                           the name of the query is taken");
     System.err.println( "       <query name>        name of the query which should be executed" );
     System.err.println( );
   }   
}
